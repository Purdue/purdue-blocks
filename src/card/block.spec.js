// card test
jest.setTimeout(30000)
import {
    createNewPost,
    enablePageDialogAccept,
    getEditedPostContent,
    insertBlock,
} from '@wordpress/e2e-test-utils';

import {
    clickElementByText,
    blockStartup,
    clickRadio,
    clickCheckbox
} from '../test-helpers'

const block = {blockTitle: 'Custom Card', blockName: 'purdue-blocks/card'}

describe( 'ðŸ”¬ Card Block', () => {
    beforeAll( async () => {
        await enablePageDialogAccept();
    } );
    beforeEach( async () => {
        await createNewPost();
    } );

    test( 'ðŸ”¬ Block should be available.', async () => {
        await insertBlock( 'Custom Card' )

        // tests that the block is properly inserted and matches the existing snapshot
        expect(await page.$('[data-type="purdue-blocks/card"]')).not.toBeNull()
        expect( await getEditedPostContent()).toMatchSnapshot()
    } )

    describe( 'ðŸ”¬ Block Editor Fields', () => {
        test( 'ðŸ”¬ Button should open Media Library.', async () => {
            await blockStartup(block)

            // open media library
            await clickElementByText('button', 'Select an image (optional)')

            // tests that media library does open
            expect(await page.$(`div.media-frame-title`)).not.toBeNull()

            /*
            NOTE: explore further to see how media items can be spoofed so that the actual
                image urls can be properly snapshotted and tested.
            */
        })
        test( 'ðŸ”¬ Title Richtext editor should be editable.', async () => {
            await blockStartup(block)

            const typeString = "Test Title"
            
            // focus the text box then type into it with the virtual keyboard
            await page.focus( '.block-editor-rich-text__editable.title' );
            await page.keyboard.type(typeString, {delay: 10})

            const editedContent = await getEditedPostContent()

            // tests: 
            // a. save output properly contains the correct string 
            // b. save output matches the existing snapshot
            expect( editedContent.includes(`<p class="title">${typeString}</p>`)).toBe(true)
            expect( editedContent).toMatchSnapshot()
        })

        test( 'ðŸ”¬ Text Richtext editor should be editable.', async () => {
            await blockStartup(block)

            const typeString = "Test text here."
            
            // focus the textarea then type into it with the virtual keyboard
            await page.focus( '.block-editor-rich-text__editable.content' );
            await page.keyboard.type(typeString, {delay: 10})

            const editedContent = await getEditedPostContent()

            // tests: 
            // a. save output properly contains the correct string 
            // b. save output matches the existing snapshot
            expect( editedContent.includes(`<p class="content">${typeString}</p>`)).toBe(true)
            expect( editedContent).toMatchSnapshot()
        })
    })
    describe( 'ðŸ”¬ Side Panel Settings', () => {
        test( 'ðŸ”¬ Gold background color should be selectable.', async () => {
            await blockStartup(block)

            // select the radio button with the virtual keyboard
            await clickRadio('Background Color', 'gold')

            const editedContent = await getEditedPostContent()

            expect( editedContent.includes(`"backgroundColor":"white"`)).toBe(false)
            expect( editedContent).toMatchSnapshot()
        })
        test( 'ðŸ”¬ Black border color should be selectable.', async () => {
            await blockStartup(block)
        
            // select the radio button with the virtual keyboard
            await clickRadio('Border Color', 'black')

            const editedContent = await getEditedPostContent()

            expect( editedContent.includes(`"borderColor":"gold"`)).toBe(false)
            expect( editedContent).toMatchSnapshot()
        })
        test( 'ðŸ”¬ Image alt text should be editable.', async () => {
            await blockStartup(block)

            const typeString = "Image alt text."
            
            // focus the text box then type into it with the virtual keyboard
            await clickElementByText('label', 'Image Alt Text')
            await page.keyboard.type(typeString, {delay: 10})

            const editedContent = await getEditedPostContent()

            // tests: 
            // a. save output properly contains the correct string 
            // b. save output matches the existing snapshot
            expect( editedContent.includes(`"altText":"${typeString}"`)).toBe(true)
            // expect( editedContent).toMatchSnapshot()
        })
        describe( 'ðŸ”¬ Link can be added to the card', () => {
            test( 'ðŸ”¬ "Add a link to this card?" checkbox reveals link inputs', async () => {
                await blockStartup(block)

                await clickCheckbox('Add a link to this card?')
                      
                let editedContent = await getEditedPostContent()
                
                expect( editedContent.includes('{"hasLink":true}')).toBe(true)
                expect( await getEditedPostContent() ).toMatchSnapshot()
                
        
                // click to uncheck the checkbox, outline should now turn off
                await clickCheckbox('Add a link to this card?')
                
                editedContent = await getEditedPostContent()
                
                expect( editedContent.includes('{"hasLink":true}')).toBe(false)
                expect( await getEditedPostContent() ).toMatchSnapshot()
            })
            test('ðŸ”¬ Link text field updates output correctly', async () => {
                await blockStartup(block)

                const typeString = "CTA text"
                await clickCheckbox('Add a link to this card?')
                await clickElementByText('label', 'Call to action text')
                await page.keyboard.type(typeString, {delay: 10})

                const editedContent = await getEditedPostContent()
                expect( editedContent.includes(`<div class="read-more-button"><span>${typeString}</span></div>`)).toBe(true)
                expect( await getEditedPostContent() ).toMatchSnapshot()
            })
            test('ðŸ”¬ Link url field updates output correctly', async () => {
                await blockStartup(block)
                await clickCheckbox('Add a link to this card?')
                const typeString = "https://www.purdue.edu"

                await clickElementByText('label', 'Link address')
                await page.keyboard.type(typeString, {delay: 10})

                const editedContent = await getEditedPostContent()
                expect( editedContent.includes(`<a href="${typeString}" target="_self" class="square-card" rel="noopener noreferrer"></a>`)).toBe(true)
                expect( await getEditedPostContent() ).toMatchSnapshot()
            })
            test('ðŸ”¬ Link can be opened in new tab', async () => {
                await blockStartup(block)

                const typeString = "https://www.purdue.edu"
                await clickCheckbox('Add a link to this card?')
                await clickElementByText('label', 'Link address')
                await page.keyboard.type(typeString, {delay: 10})

                await clickCheckbox('Open link in new tab?')

                const editedContent = await getEditedPostContent()
                expect( editedContent.includes(`<a href="${typeString}" target="_blank" class="square-card" rel="noopener noreferrer"></a>`)).toBe(true)
                expect( await getEditedPostContent() ).toMatchSnapshot()
            })
        })
    })
} );
