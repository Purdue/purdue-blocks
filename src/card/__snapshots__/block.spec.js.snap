// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`🔬 Card Block 🔬 Block Editor Fields 🔎 Text Richtext editor should be editable. 1`] = `
"<!-- wp:purdue-blocks/card -->
<div class=\\"square-card\\"><p class=\\"content\\">Test text here.</p></div>
<!-- /wp:purdue-blocks/card -->"
`;

exports[`🔬 Card Block 🔬 Block Editor Fields 🔎 Title Richtext editor should be editable. 1`] = `
"<!-- wp:purdue-blocks/card -->
<div class=\\"square-card\\"><p class=\\"title\\">Test Title</p></div>
<!-- /wp:purdue-blocks/card -->"
`;

exports[`🔬 Card Block 🔬 Block should be available. 1`] = `
"<!-- wp:purdue-blocks/card -->
<div class=\\"square-card\\"></div>
<!-- /wp:purdue-blocks/card -->"
`;

exports[`🔬 Card Block 🔬 Side Panel Settings 🔎 Black border color should be selectable. 1`] = `
"<!-- wp:purdue-blocks/card {\\"borderColor\\":\\"black\\"} -->
<div class=\\"square-card border-black\\"></div>
<!-- /wp:purdue-blocks/card -->"
`;

exports[`🔬 Card Block 🔬 Side Panel Settings 🔎 Gold background color should be selectable. 1`] = `
"<!-- wp:purdue-blocks/card {\\"backgroundColor\\":\\"gold\\"} -->
<div class=\\"square-card background-gold\\"></div>
<!-- /wp:purdue-blocks/card -->"
`;

exports[`🔬 Card Block 🔬 Side Panel Settings 🔬 Link can be added to the card 🔎 "Add a link to this card?" checkbox reveals link inputs 1`] = `
"<!-- wp:purdue-blocks/card {\\"hasLink\\":true} -->
<a href=\\"\\" target=\\"_self\\" class=\\"square-card\\" rel=\\"noopener noreferrer\\"></a>
<!-- /wp:purdue-blocks/card -->"
`;

exports[`🔬 Card Block 🔬 Side Panel Settings 🔬 Link can be added to the card 🔎 "Add a link to this card?" checkbox reveals link inputs 2`] = `
"<!-- wp:purdue-blocks/card -->
<div class=\\"square-card\\"></div>
<!-- /wp:purdue-blocks/card -->"
`;

exports[`🔬 Card Block 🔬 Side Panel Settings 🔬 Link can be added to the card 🔎 Link can be opened in new tab 1`] = `
"<!-- wp:purdue-blocks/card {\\"hasLink\\":true,\\"link\\":\\"https://www.purdue.edu\\",\\"external\\":true} -->
<a href=\\"https://www.purdue.edu\\" target=\\"_blank\\" class=\\"square-card\\" rel=\\"noopener noreferrer\\"></a>
<!-- /wp:purdue-blocks/card -->"
`;

exports[`🔬 Card Block 🔬 Side Panel Settings 🔬 Link can be added to the card 🔎 Link text field updates output correctly 1`] = `
"<!-- wp:purdue-blocks/card {\\"hasLink\\":true,\\"linkText\\":\\"CTA text\\"} -->
<a href=\\"\\" target=\\"_self\\" class=\\"square-card\\" rel=\\"noopener noreferrer\\"><div class=\\"read-more-button\\"><span>CTA text</span></div></a>
<!-- /wp:purdue-blocks/card -->"
`;

exports[`🔬 Card Block 🔬 Side Panel Settings 🔬 Link can be added to the card 🔎 Link url field updates output correctly 1`] = `
"<!-- wp:purdue-blocks/card {\\"hasLink\\":true,\\"link\\":\\"https://www.purdue.edu\\"} -->
<a href=\\"https://www.purdue.edu\\" target=\\"_self\\" class=\\"square-card\\" rel=\\"noopener noreferrer\\"></a>
<!-- /wp:purdue-blocks/card -->"
`;
